Add metric store and registry types

simonjbeaumont opened the pull request:
## Motivation

As part of the effort to provide an OLTP backend for Swift Metrics (see #84), we need some metric types and a registry. Since the intention is that these types support only the Swift Metics API (cf. the full OTel Metrics API), there's nothing that needs to be unique about them and they can be taken wholesale from another Swift Metrics backend package that has a good storage implementation, namely Swift Prometheus.

## Modifications

- Pull in the core metrics (Counter, Gauge, Histogram) and registry types from Swift Prometheus.
- Remove the Prometheus-specific emitting functionality.
- Add a NOTICE file that credits Swift Prometheus for these types.

## Result

Core metrics types exist, which can be wired up in either direction, both of which will be in subsequent pull requests:
1. To provide a Swift Metrics backend.
2. To export via OTLP.

slashmo commented on NOTICE.txt:
```diff
@@ -0,0 +1,57 @@
+
+  * https://github.com/slashmo/swift-otel
+
+Copyright 2023 The Swift OTel Project
```
```suggestion
Copyright 2024 The Swift OTel Project
```

simonjbeaumont commented:
@slashmo At this point I think I've done all I can to appease the code coverage CI as the rest are `fatalError` paths (lifted from Swift Prometheus) for when you try to register a metric that has already been registered with the registry.

slashmo commented:
> At this point I think I've done all I can to appease the code coverage CI as the rest are fatalError paths (lifted from Swift Prometheus) for when you try to register a metric that has already been registered with the registry.
> @simonjbeaumont 

Thanks for adding all these tests 🙏 As discussed offline, do we want to eliminate some of these fatal error cases to get closer to the OTel spec? Perhaps then I wouldn't need to change the coverage thresholds. If not, I'm also okay with adjusting them accordingly.

simonjbeaumont commented:
OK, so I've revisited the requirements for the metrics registry and now accommodate what the OTel specification says. Specifically, when registering duplicate instruments, it will warn and still provide a functional metric: https://opentelemetry.io/docs/specs/otel/metrics/sdk/#duplicate-instrument-registration.

I've made this functionality customisable. You can instantiate the registry using `OTelMetricRegistry(onDuplicateRegistration: .crash)` if you prefer to crash hard. But the default will emit a warning log message, as suggested by the OTel spec.

And, in order to satisfy the unyielding 100% coverage tests, I have added hooked fatal error tests for testing the crashing duplicate registration handler.

(I've also updated my WIP branch with the other layers too, if you needed to reference how this composes up the stack).

simonjbeaumont commented:
```
[603/658] Compiling OTelTestingTests HookedFatalErrorTests.swift
/home/runner/work/swift-otel/swift-otel/Tests/OTelTestingTests/HookedFatalErrorTests.swift:58:9: error: cannot find 'XCTExpectFailure' in scope
        XCTExpectFailure {
        ^~~~~~~~~~~~~~~~
```

OK, so I forgot that Linux doesn't have `XCTExpectFailure`...

simonjbeaumont commented:
OK, so I moved some of the functionality from the test target to the `OTelTesting` target, which seems to have tripped up the coverage again. This seems to be the wrong incentive, since, based on what I saw, you wanted reusable test utilities to be in `OTelTesting` and I had a need for them in the tests in upcoming PRs too, so wanted to factor out. 

Potentially OTelTesting should be removed from your coverage report?

slashmo commented on Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift:
```diff
@@ -0,0 +1,284 @@
+    func unregisterCounter(_ counter: Counter) {
+        let identifier = counter.instrumentIdentifier
+        self.storage.withLockedValue { storage in
+            if var existingInstrument = storage.counters[identifier] {
+                existingInstrument.removeValue(forKey: counter.attributes)
+                if existingInstrument.isEmpty {
+                    storage.counters.removeValue(forKey: identifier)
+                    storage.unregister(identifier, forName: identifier.name)
+                }
+            }
```
This pattern of removing and clearing the storage is used in all `unregister` methods, as far as I can tell. Do you think it's worth refactoring into a method?

Something like:

```swift
private func unregister<Instrument: IdentifiableInstrument>(_ keyPath: KeyPath<Storage, [InstrumentIdentifier: Instrument]>) {
    let identifier = instrument.identifier
    storage.withLockedValue { storage in
        if var existingInstrument = storage[keyPath: keyPath][identifier] {
            existingInstrument.removeValue(forKey: instrument.attributes)
            if existingInstrument.isEmpty {
                storage[keyPath: keyPath].removeValue(forKey: identifier)
                storage.unregister(identifier, forName: identifier.name)
            }
        }
    }
}

func unregisterCounter(_ counter: Counter) {
    self.unregister(\.counters, counter)
}
```

This would require adding a new protocol or extending `IdentifiableInstrument` to also require `attributes`. Please note that I haven't compiled this code snippet, it's just a brain dump.

simonjbeaumont replied:
IMHO I don't think this is worth it. I _did_ factor out the duplicate registration logic as it is independent of the instrument kind, but that needed to be pushed into the storage type too because we're holding the lock.

I'm happy to revisit this as an implementation detail if we're happy with the API, so we don't block this PR, which is holding up the others. WDYT?

slashmo replied:
Yeah, fair point. Let's merge first and optionally look back at it in the future 👍

slashmo commented on Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift:
```diff
@@ -0,0 +1,284 @@
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>) {
+        self.logger.warning("Duplicate instrument registration", metadata: [
+            "newRegistration": "\(newRegistration)",
+            "existingRegistrations": "\(existingRegistrations)",
```
```suggestion
            "existingRegistrations": .array(existingRegistrations.map { "\($0)" }),
```

Nitpicky, but I think this might read a bit nicer in some log handlers.

simonjbeaumont replied:
Sure, done.

slashmo commented on Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift:
```diff
@@ -0,0 +1,284 @@
+    let logger: Logger
+
+    func handle(newRegistration: InstrumentIdentifier, existingRegistrations: Set<InstrumentIdentifier>) {
+        self.logger.warning("Duplicate instrument registration", metadata: [
```
Should this be limited to only spitting out a warning once per running instance? Otherwise, it might accidentally lead to flooding the user's logging system. Maybe @ktoso can give some input on this, I know you're a fan of this warn-only-once approach.

slashmo replied:
> It currently does it only once per duplicate instrument, which is the right thing to do, both in terms of the spec and for usability. It would be super annoying, if you had a few duplicate instruments to have to fix them one by one and re-run. That would be like your unit test suite stopping after the first failed test case.

Oh, I must have missed that part when reading the PR. One warning per instrument sounds good to me 👍

> We could consider adding a complete opt-out `.none` value for the `DuplicateRegistrationBehavior`, but I think we're pushing people in the wrong direction.

Agreed, I think we should hold off on doing it until requested by users.

> Also worth pointing out that this check and the logging is only performed after lookup has failed, which means that, if the library is used correctly (no conflicting identities) then it will take the fast path and just return the existing instrument.

Awesome 👌

slashmo commented on Sources/OTelTesting/RecordingDuplicateRegistrationHandler.swift:
```diff
@@ -0,0 +1,24 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import NIOConcurrencyHelpers
```
```suggestion

import NIOConcurrencyHelpers
```

slashmo commented on Sources/OTelTesting/RecordingLogHandler.swift:
```diff
@@ -0,0 +1,41 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+import Logging
```
```suggestion

import Logging
```

slashmo replied:
Thanks. Good point, I created an issue for adding this: https://github.com/slashmo/swift-otel/issues/87

slashmo commented on Sources/OTelTesting/XCTAssertThrowsFatalError.swift:
```diff
@@ -0,0 +1,37 @@
+import OTel
+import XCTest
+
+package func XCTAssertThrowsFatalError(
```
Nice! 👍

slashmo commented on Tests/OTelTestingTests/HookedFatalErrorTests.swift:
```diff
@@ -0,0 +1,56 @@
+@testable import OTelTesting
+import XCTest
+
+final class HookedFatalErrorTests: XCTestCase {
```
I appreciate you adding this extra test target to increase code coverage, but I don't think we necessarily need it anymore. What do you think, now that `OTelTesting` is code-coverage-ignored anyways? I'm happy to merge this as is if you think it's still valuable to have this test case and given you already implemented the tests.

slashmo commented on Tests/OTelTests/Metrics/MetricStore/CounterTests.swift:
```diff
@@ -0,0 +1,67 @@
+import XCTest
+
+final class CounterTests: XCTestCase {
+    func testIncrement() {
```
I'm happy to perform these changes since they'd be very cumbersome, but in the spirit of adhering to the common naming convention in the library, could we please name these new test methods accordingly? Here, `test_increment` would be more in line with other existing test methods.

slashmo commented on Tests/OTelTests/Metrics/MetricStore/CounterTests.swift:
```diff
@@ -0,0 +1,67 @@
+        counter.assertStateEquals(integerPart: 0, doublePart: 0)
+    }
+
+    func testConcurrentIncrement() async {
```
Regarding the naming suggestion, this could be `test_increment_concurrently`.

slashmo reviewed (approved):
Thanks a lot for the various iterations on this PR 🙏 I think it now strikes a nice balance between using types from `swift-prometheus` and still behaving similarly to what OTel users would expect in terms of handling duplicates.

I added a few refactoring suggestions and nitpicks inline but am very happy with the overall shape of these changes 👍

simonjbeaumont commented on Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift:
```diff
@@ -0,0 +1,284 @@
+import Logging
+import struct NIOConcurrencyHelpers.NIOLockedValueBox
+
+public final class OTelMetricRegistry: Sendable {
```
Sure, I've added a first draft. FWIW, I was expecting that we'd do a public API review pass before cutting a release, and make sure we polish the docs etc.

Same goes for the other public symbols: added some docs; happy to polish some more in a follow up.

FWIW, I think swift-format has a rule to check for missing documentation comments on public symbols,.

simonjbeaumont replied:
On the subject of public API, we should also discuss whether we _want_ this to be public or not. It's currently the point of configuration and is passed to the `OTLPMetricsFactory` (see the WIP branch), but we could make an argument for making this entirely internal and exposing the config on the `OTLPMetricsFactory` itself.

For now, I think we leave it as is, and once we have all the pieces we can work out how much API surface we want to expose?

slashmo replied:
Yeah, that's a good plan. Thanks for adding some documentation now, agreed that we should revisit it for the complete project later on.

simonjbeaumont commented on Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift:
```diff
@@ -0,0 +1,284 @@
+        self.storage = .init(Storage(duplicateRegistrationHandler: duplicateRegistrationHandler))
+    }
+
+    public convenience init(onDuplicateRegistration: DuplicateRegistrationBehavior = .warn) {
```
Done. I didn't spell out that it defaults to `.warn` since that's the default parameter. Usually you only document the the default when the parameter is an optional that defaults to `nil` and you want to explain what happens when you use `nil`.

slashmo replied:
Thanks, makes sense 👍

simonjbeaumont commented on Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift:
```diff
@@ -0,0 +1,284 @@
+
+        var behavior: Behavior
+        public static let warn = Self(behavior: .warn)
+        public static let crash = Self(behavior: .crash)
```
Basic docs added.

simonjbeaumont commented on Sources/OTel/Metrics/MetricStore/Counter.swift:
```diff
@@ -0,0 +1,81 @@
+
+    func increment(by amount: Double) {
+        precondition(amount >= 0)
+        // We busy loop here until we can update the atomic successfully.
+        // Using relaxed ordering here is sufficient, since the as-if rules guarantess that
+        // the following operations are executed in the order presented here. Every statement
+        // depends on the execution before.
+        while true {
```
Technically we could. Shall we push it into a follow up to unblock the other PRs?

simonjbeaumont reviewed (commented):
@slashmo I've responded to all your comments and actioned almost all of them.

slashmo commented on Sources/OTel/Metrics/MetricStore/OTelMetricRegistry.swift:
```diff
@@ -0,0 +1,300 @@
+
+    /// Create a new ``OTelMetricRegistry``.
+    /// - Parameters:
+    ///   - onDuplicateRegistration: Behavior when when more than one instrument of the same
```
```suggestion
    ///   - onDuplicateRegistration: Behavior when more than one instrument of the same
```
